cdap-app-fabric/src/main/java/org/apache/twill/internal/yarn/AbstractYarnTwillService.java

```diff
@@ -98,4 +115,42 @@ protected final boolean handleSecureStoreUpdate(Message message) {
 
     return true;
   }
+
+  private static void cloneHaNnCredentials(Location location,
+                                           UserGroupInformation ugi) throws URISyntaxException, IOException {
+    Configuration hConf = getConfiguration(location.getLocationFactory());
+
+    Map<String, Map<String, InetSocketAddress>> nsIdMap = DFSUtil.getHaNnRpcAddresses(hConf);
+    for (Map.Entry<String, Map<String, InetSocketAddress>> entry : nsIdMap.entrySet()) {
+      String nsId = entry.getKey();
+      Map<String, InetSocketAddress> addressesInNN = entry.getValue();
+      if (!HAUtil.isHAEnabled(hConf, nsId) || addressesInNN == null || addressesInNN.isEmpty()) {
+        continue;
+      }
+      // The client may have a delegation token set for the logical
+      // URI of the cluster. Clone this token to apply to each of the
+      // underlying IPC addresses so that the IPC code can find it.
+      LOG.info("Cloning delegation token for nsId {}", nsId);
+      URI uri = new URI(HdfsConstants.HDFS_URI_SCHEME + "://" + nsId);
```

Comment:
nit: use ``URI.create(...)`` so you don't have to propagate the ``URISyntaxException``, since you're constructing the URI here and know it won't throw ``URISyntaxException``